### PR TITLE
feat(auth): Add support for guest tokens with a new provider

### LINKED ISSUE #12394: Allow guests to generate tokens for development
## Feature Suggestion

@freben and I discussed this offline.

When using the sign in as guest feature it would be great if tokens were generated by the fetchApi and token manager. This would reduce the friction to getting a local development Backstage up and running.

## Possible Implementation

Possibly a `allowGuestMode` flag in `app-config.local.yaml` that generates tokens from a random seed.

## Context

While setting up some new endpoints for the Kubernetes backend (https://github.com/backstage/backstage/pull/12393) it was quite difficult to get started. The new endpoints call the catalog api which needs a Backstage token, so I had to login using a local auth setup rather than just using a guest account.

This complicates the setup process and creates friction when developers are trying to explore Backstage.

### PR DESCRIPTION
**Pull request by sennyeya**
## Hey, I just made a Pull Request!

Fixes #12394. I wasn't sure what the state of #18139, @RubenV-dev PTAL at this and see if this is similar to what you had in mind.

The solution I landed on here is a bit of mix. We don't need to point to an external provider as we're the ones emitting the token so we can skip the OAuth step for start. There's no `providerInfo` per se at the `handler/frame` step, so we just emit a default token with some override-able properties. I also added refresh capability for a better user experience in guest mode.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/app-defaults/src/defaults/apis.ts
Comment (Rugvip) [inline packages/app-defaults/src/defaults/apis.ts:264]: Tbh I think it'd be best to implement this as something that's completely built-in in the sign-in page instead, rather than having a Utility API for it. There's no point calling this API outside the sign-in page right?

File: packages/core-app-api/src/lib/AuthConnector/DirectAuthConnector.ts
Comment (Rugvip) [inline packages/core-app-api/src/lib/AuthConnector/DirectAuthConnector.ts:73]: Could this be kept as is? you don't seem to be overriding it

File: packages/core-app-api/src/apis/implementations/auth/guest/GuestAuth.ts
Comment (Rugvip) [inline packages/core-app-api/src/apis/implementations/auth/guest/GuestAuth.ts:48]: I think you could use this instead right? https://github.com/backstage/backstage/blob/b5c65ca6e9ec7f192ddbf59afc20f1096e8c724e/packages/core-components/src/layout/ProxiedSignInPage/ProxiedSignInIdentity.ts#L80

File: plugins/auth-backend/src/providers/providers.ts
Comment (Rugvip) [inline plugins/auth-backend/src/providers/providers.ts:88]: We're avoiding introducing any new built-in providers in the auth backend. It's worth a discussion whether this one is special, but for now I'd say it's best to avoid

File: plugins/auth-backend-module-guest-provider/src/createGuestAuthFactory.ts
Comment (Rugvip) [inline plugins/auth-backend-module-guest-provider/src/createGuestAuthFactory.ts:35]: I think the existing `createProxyAuthProviderFactory` should do it here tbh, and you can avoid adding all of these.

File: plugins/auth-backend-module-guest-provider/src/module.ts
Comment (Rugvip) [inline plugins/auth-backend-module-guest-provider/src/module.ts:34]: 👍, I think it's a bit too common that people forget to set this though, I think an explicit check the other way around is gonna be safer? i.e. `NODE_ENV !== 'development'`

**Review by Rugvip — CHANGES_REQUESTED** [review]
Sweet! 😁 🎉  One of those things that have been on the radar for while but never found the time 😅 

Main nit is that I think we can rely a lot more on the existing implementation and avoid a lot of the additions. Guest auth is basically proxy auth without the auth step.

File: packages/app-defaults/src/defaults/apis.ts
Comment (sennyeya) [reply packages/app-defaults/src/defaults/apis.ts]: Yup, what would be a good way to ensure people have the module installed? Currently it's all frontend and doesn't require a backend installation.

File: plugins/auth-backend-module-guest-provider/src/createGuestAuthFactory.ts
Comment (sennyeya) [reply plugins/auth-backend-module-guest-provider/src/createGuestAuthFactory.ts]: 😅 that makes this implementation a lot simpler

File: plugins/auth-backend/src/providers/providers.ts
Comment (sennyeya) [reply plugins/auth-backend/src/providers/providers.ts]: If my proposed changes to the `guestProvider` are the in the right direction, this might make sense to add to the old backend? At least until folks get transferred over to the new backend.

**Review by freben — COMMENTED** [review]
Cool stuff!

File: plugins/auth-backend-module-guest-provider/src/authenticator.ts
Comment (sennyeya) [reply plugins/auth-backend-module-guest-provider/src/authenticator.ts]: I hate when this happens, updated in https://github.com/backstage/backstage/pull/22565/commits/df1e8303252bef6372da1831ffb7011c02854a4d

File: plugins/auth-backend-module-guest-provider/README.md
Comment (sennyeya) [reply plugins/auth-backend-module-guest-provider/README.md:7]: Added in https://github.com/backstage/backstage/pull/22565/commits/df1e8303252bef6372da1831ffb7011c02854a4d

File: packages/core-components/src/layout/SignInPage/guestProvider.tsx
Comment (freben) [inline packages/core-components/src/layout/SignInPage/guestProvider.tsx:71]: Do we need these lines at all? Spontaneously it might be ok to only call start, or to get an identity, not both?

File: packages/core-components/src/layout/SignInPage/guestProvider.tsx
Comment (freben) [inline packages/core-components/src/layout/SignInPage/guestProvider.tsx:43]: So, if we put this code here, everybody who's on the old backend system will no longer be able to use guest auth.

Could you make it so that it first tries the proxy auth, and if it throws / doesn't return an identity (see the comment in `loader`), it opens a popup (even a silly `confirm()`) asking the user "The guest auth provider is not installed, do you want to continue with a classic token-less guest identity?"

If you do so, you may want to like flag in local storage or whatnot, what the user choice was, because the loader will need to know next time around which one of the identities to try. Essentially if the user responded yes, it'll just return GuestUserIdentity like before.

**Review by freben — COMMENTED** [review]
Alright sorry, finally got back to this again!

File: packages/core-components/src/layout/SignInPage/guestProvider.tsx
Comment (aramissennyeydd) [reply packages/core-components/src/layout/SignInPage/guestProvider.tsx]: Yup, good call. Given that _most_ installations probably have this set up in one form or another, I think backporting this module to the old backend wouldn't fix this issue as I was originally thinking. In that vein, I've updated this to now check if you have the new backend module installed, if yes, we use the proxied sign in. If not, we show a confirm and then store the positive result in local storage. If they choose the negative option, "cancel" here, we just trigger a signin failure, as it's an impossible to use Backstage instance with no identity.

Then on load, we check if a) they have the backend module set up, if so we use proxy auth, b) if they have the guest override set _and_ the new backend module, if so we try to get them to choose one and  will not remember their choice (we can change this, I'm not 100% why people would want to continue using the old frontend-only tokens), c) if they have the guest override set and no valid identity from the backend module (either not installed or not configured correctly), we use the old way of authenticating. 

This does mean we're now making a single API request per load in the guest account, but I think the experience for folks wanting to transfer over to the new module will be better this way. They won't have to dig around in their local storage to remove items.

File: packages/core-components/src/layout/SignInPage/guestProvider.tsx
Comment (aramissennyeydd) [reply packages/core-components/src/layout/SignInPage/guestProvider.tsx]: Updated, see above comment.

File: plugins/auth-backend-module-guest-provider/src/resolvers.ts
Comment (aramissennyeydd) [reply plugins/auth-backend-module-guest-provider/src/resolvers.ts]: I like `user:development/guest`, I think there's a longer RFC for updating that globally, but this could be a good module to enforce that default and start moving folks towards it.

The auth context is a little weird here, either we allow configuring this at the `auth.*` level or the `auth.providers.guest.*` level. I'm open to either, but if we get to a future where we allow this to be easily configurable for the guest module, would we want this to be configurable across the application? Are there any benefits for that, ie from an auth perspective?

**Review by freben — APPROVED** [review]
Alright I'm happy now, over to @Rugvip !

File: plugins/auth-backend-module-guest-provider/src/resolvers.ts
Comment (aramissennyeydd) [reply plugins/auth-backend-module-guest-provider/src/resolvers.ts:27]: Something like our Uffizzi instances would complain though right? This might be useful for CI environments... ? Up to you/the other maintainers, I'm good either way.

**Comment by freben** [comment]
@aramissennyeydd sorry this needs a rebase

File: packages/core-components/src/layout/SignInPage/guestProvider.tsx
Comment (Rugvip) [inline packages/core-components/src/layout/SignInPage/guestProvider.tsx:36]:
```suggestion
      error.name === 'ResponseError' &&
      (error as ResponseError).cause.name === 'NotFoundError'
```

`instanceof` is dangerous, since you don't know that the `identity` implementation uses the same version of `@backstage/errors` as this package. This check is kinda the best we can do for now, but a more thorough `ResponseError.is(...)` or something like that might be nice

**Comment by aramissennyeydd** [comment]
@Rugvip Not sure if this is related to the new auth work, but I'm not able to test any changes to this work locally. Getting a 401 Unauthorized due to a call in this block, https://github.com/sennyeya/backstage/blob/4b277033714facd9144c9ea93ed78b0a3114f812/plugins/auth-backend/src/lib/resolvers/CatalogAuthResolverContext.ts#L148-L159

It looked like we're bypassing this for Github auth by providing a custom resolver implementation that doesn't call the catalog (?).

Full error below,
```
{
    "error": {
        "name": "ResponseError",
        "message": "Request failed with 401 Unauthorized",
        "response": {
            "size": 0,
            "timeout": 0
        },
        "body": {
            "error": {
                "name": "ResponseError",
                "message": "Request failed with 401 Unauthorized",
                "response": {
                    "size": 0,
                    "timeout": 0
                },
                "body": {
                    "error": {
                        "name": "AuthenticationError",
                        "message": "\"alg\" (Algorithm) Header Parameter value not allowed",
                        "stack": "AuthenticationError: \"alg\" (Algorithm) Header Parameter value not allowed\n    at DefaultIdentityClient.getIdentity (/backstage/plugins/auth-node/src/identity/DefaultIdentityClient.ts:90:13)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at <anonymous> (/backstage/plugins/permission-backend/src/service/router.ts:191:20)"
                    },
                    "request": {
                        "method": "POST",
                        "url": "/authorize"
                    },
                    "response": {
                        "statusCode": 401
                    }
                },
                "cause": {
                    "name": "AuthenticationError",
                    "message": "\"alg\" (Algorithm) Header Parameter value not allowed",
                    "stack": "AuthenticationError: \"alg\" (Algorithm) Header Parameter value not allowed\n    at DefaultIdentityClient.getIdentity (/backstage/plugins/auth-node/src/identity/DefaultIdentityClient.ts:90:13)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at <anonymous> (/backstage/plugins/permission-backend/src/service/router.ts:191:20)"
                },
                "statusCode": 401,
                "statusText": "Unauthorized",
                "stack": "ResponseError: Request failed with 401 Unauthorized\n    at Function.fromResponse (/backstage/packages/errors/src/errors/ResponseError.ts:76:12)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at PermissionClient.makeRequest (/backstage/plugins/permission-common/src/PermissionClient.ts:160:13)\n    at AuthorizedEntitiesCatalog.entities (/backstage/plugins/catalog-backend/src/service/AuthorizedEntitiesCatalog.ts:54:7)\n    at <anonymous> (/backstage/plugins/catalog-backend/src/service/createRouter.ts:192:30)"
            },
            "request": {
                "method": "GET",
                "url": "/entities/by-name/user/development/guest"
            },
            "response": {
                "statusCode": 401
            }
        },
        "cause": {
            "response": {
                "size": 0,
                "timeout": 0
            },
            "body": {
                "error": {
                    "name": "AuthenticationError",
                    "message": "\"alg\" (Algorithm) Header Parameter value not allowed",
                    "stack": "AuthenticationError: \"alg\" (Algorithm) Header Parameter value not allowed\n    at DefaultIdentityClient.getIdentity (/backstage/plugins/auth-node/src/identity/DefaultIdentityClient.ts:90:13)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at <anonymous> (/backstage/plugins/permission-backend/src/service/router.ts:191:20)"
                },
                "request": {
                    "method": "POST",
                    "url": "/authorize"
                },
                "response": {
                    "statusCode": 401
                }
            },
            "cause": {
                "name": "AuthenticationError",
                "message": "\"alg\" (Algorithm) Header Parameter value not allowed",
                "stack": "AuthenticationError: \"alg\" (Algorithm) Header Parameter value not allowed\n    at DefaultIdentityClient.getIdentity (/backstage/plugins/auth-node/src/identity/DefaultIdentityClient.ts:90:13)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at <anonymous> (/backstage/plugins/permission-backend/src/service/router.ts:191:20)"
            },
            "statusCode": 401,
            "statusText": "Unauthorized",
            "name": "ResponseError",
            "message": "Request failed with 401 Unauthorized",
            "stack": "ResponseError: Request failed with 401 Unauthorized\n    at Function.fromResponse (/backstage/packages/errors/src/errors/ResponseError.ts:76:12)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at PermissionClient.makeRequest (/backstage/plugins/permission-common/src/PermissionClient.ts:160:13)\n    at AuthorizedEntitiesCatalog.entities (/backstage/plugins/catalog-backend/src/service/AuthorizedEntitiesCatalog.ts:54:7)\n    at <anonymous> (/backstage/plugins/catalog-backend/src/service/createRouter.ts:192:30)"
        },
        "statusCode": 401,
        "statusText": "Unauthorized",
        "stack": "ResponseError: Request failed with 401 Unauthorized\n    at Function.fromResponse (/backstage/packages/errors/src/errors/ResponseError.ts:76:12)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at CatalogClient.requestOptional (/backstage/packages/catalog-client/src/CatalogClient.ts:449:13)\n    at CatalogAuthResolverContext.findCatalogUser (/backstage/plugins/auth-backend/src/lib/resolvers/CatalogAuthResolverContext.ts:115:16)\n    at CatalogAuthResolverContext.signInWithCatalogUser (/backstage/plugins/auth-backend/src/lib/resolvers/CatalogAuthResolverContext.ts:149:24)\n    at Object.refresh (/backstage/plugins/auth-node/src/proxy/createProxyRouteHandlers.ts:66:24)"
    },
    "request": {
        "method": "GET",
        "url": "/api/auth/guest/refresh"
    },
    "response": {
        "statusCode": 401
    }
}
```

**Comment by Rugvip** [comment]
@aramissennyeydd very likely related! It's prolly the new permission client change tbh. Will have a look later today or tomorrow before the release unless you get to it first!

**Comment by aramissennyeydd** [comment]
@Rugvip Looks like we have 2 token manager instances ([manager 1 from auth](https://github.com/sennyeya/backstage/blob/5d1046dd206e1f264120a2ff28ef5acb89e8c3c3/packages/backend-app-api/src/services/implementations/auth/authServiceFactory.ts#L210-L212) and [manager 2 for permissions](https://github.com/sennyeya/backstage/blob/5d1046dd206e1f264120a2ff28ef5acb89e8c3c3/packages/backend-app-api/src/services/implementations/permissions/permissionsServiceFactory.ts#L30)) and since the verification keys are stored in memory, requests to one token manager aren't valid in the second manager in dev mode with no secrets.

Should be a straightforward fix, but I'll defer to you on the right path forward + PR.

**Comment by Rugvip** [comment]
@aramissennyeydd thank you! 👍 It's indeed the issue, although probably not those two exact examples. Seems like the catalog not being migrated yet highlighted this issue, which is good 😁

**Review by Rugvip — APPROVED** [review]
Awesome, let's :shipit:! 🎉 

Tried it out a bit locally, works like a charm 👌 np switching between old and new system etc. 👍

**Pull request merged**
